src/storage-controller/src/lib.rs

```diff
@@ -442,6 +442,81 @@ where
             .drop_replica(replica_id);
     }
 
+    async fn evolve_nullability_for_bootstrap(
+        &mut self,
+        storage_metadata: &StorageMetadata,
+        collections: Vec<(GlobalId, RelationDesc)>,
+    ) -> Result<(), StorageError<Self::Timestamp>> {
+        let persist_client = self
+            .persist
+            .open(self.persist_location.clone())
+            .await
+            .unwrap();
+
+        for (global_id, relation_desc) in collections {
+            let shard_id = storage_metadata.get_collection_shard(global_id)?;
+            let diagnostics = Diagnostics {
+                shard_name: global_id.to_string(),
+                handle_purpose: "evolve nullability for bootstrap".to_string(),
+            };
+            let latest_schema = persist_client
+                .latest_schema::<SourceData, (), T, Diff>(shard_id, diagnostics)
+                .await
+                .expect("invalid persist usage");
+            let Some((schema_id, existing_relation_desc, _)) = latest_schema else {
+                continue;
+            };
+
+            match relation_desc.diff(&existing_relation_desc) {
```

Comment:
Can you expand on why we need to add this additional check? (I would have thought that Persist-level compatibility was both necessary and sufficient.)

Reply:
We do this additional check because AFAICT if the the existing `compare_and_evolve_schema` method in Persist doesn't no-op if the schema is exactly the same. So if we didn't do this check we would be registering a new schema on every startup.

Maybe a better way to handle this though is to have Persist's `compare_and_evolve_schema` no-op if the schema is already exactly the same?

Reply:
> Maybe a better way to handle this though is to have Persist's compare_and_evolve_schema no-op if the schema is already exactly the same?
Seems totally reasonable IMO. Probably just an oversight on my part when typing it up

Reply:
Sweet! Just pushed that up in the latest commit, will think through a bit more but I like relying only on Persist's schema evolution

Reply:
Great, thanks! Making that call idempotent definitely feels much less error prone to me.